src/battle_ai_main.c

```diff
@@ -3564,11 +3564,17 @@ static u32 AI_CalcMoveEffectScore(u32 battlerAtk, u32 battlerDef, u32 move)
         IncreaseParalyzeScore(battlerAtk, battlerDef, move, &score);
         break;
     case EFFECT_SUBSTITUTE:
-        ADJUST_SCORE(GOOD_EFFECT);
+        if (HasAnyKnownMove(battlerDef) && GetBestDmgFromBattler(battlerDef, battlerAtk) < gBattleMons[battlerAtk].maxHP / 4)
```

Comment:
what's the purpose of `HasAnyKnownMove`?

Reply:
Without that check, on turn 1 of a brand new battle, the AI won't know any of the player's moves, so `GetBestDmgFromBattler` will always return 0 and the AI will always be incentivized to use Substitute turn 1 in all cases.

With that check, it makes sure that the AI has determined whether or not Substitute is a good idea based off seeing at least one move first.

Reply:
Could you replace the In the function `GetBestDmgFromBattler` `u32 bestDmg = 0` with `u32 bestDmg = UINT32_MAX`? You could also make a define for `UINT32_MAX` like `UNKNOWN_DAMAGE`. With this you shouldn't need the function anymore.

Reply:
Sure! :)

Reply:
Actually doesn't this not work because `AI_DATA->simulatedDmg[battler][battlerTarget][i].expected` will never be bigger than UINT32_MAX?

Reply:
oh right haven't thought this one through. then I would check if damage is 0 and make store the bestdamage result in a var to avoid calling it twice. Also only if you want you could do something like:
```
ADJUST_DAMAGE(CalcSubstituteScore(...))
```
and return in `CalcSubstituteScore(...))` a `tempScore` but very optional

Reply:
Macros make me nervous I will leave that for a future person to refactor if they like :)

Will rework the function a bit and let you know when ready!

Reply:
you aren't adding a new macro. It is just the functiopn

Reply:
How's this
```
u32 GetBestDmgFromBattler(u32 battler, u32 battlerTarget)
{
    u32 i;
    u32 bestDmg = UINT32_MAX, moveDmg = 0;
    u32 unusable = AI_DATA->moveLimitations[battler];
    u16 *moves = GetMovesArray(battler);

    for (i = 0; i < MAX_MON_MOVES; i++)
    {
        if (moves[i] != MOVE_NONE && moves[i] != MOVE_UNAVAILABLE && !(unusable & gBitTable[i]))
        {
            moveDmg = AI_DATA->simulatedDmg[battler][battlerTarget][i].expected;
            if (moveDmg > bestDmg || bestDmg = UINT32_MAX)
                bestDmg = moveDmg;
        }
    }

    return bestDmg;
}
```
This `if (moveDmg > bestDmg || bestDmg == UINT32_MAX)` is a bit weird but it works for our purposes for now at least

Reply:
I don't feel comfortable editing the function for a purpose. It should stay universal.

What my idea was in the substitute case :
```
u32 bestDamge = GetBestDmgFromBattler(battlerDef, battlerAtk)
```
and then 
```
if (bestDamage != 0 && ..)
```
This avoids having to check for move none (also it didn't really solve the issue because you would need to check if it is a damage move that can deal damage)

Reply:
Forgive me if I'm just being dense, but doesn't that just put us in the same position as before, where because we're checking `if (bestDamage != 0 && ..)` we'll never use Substitute against mons that only have non damaging moves

I'm sorry I'm having a hard time grasping this lol

Reply:
ye, that's true. 

But on an other note. Aren't we trying to avoid the issue that the AI will not know any moves from player? The correct way would be either to come up with alternative logic to use the sub or except that when it doesn't know the players move it will use sub on the first turn always.  

Basically we are trying to make the AI cheat without cheating. Not sure if that makes sense.

Reply:
I'm trying to avoid the issue of "If the AI doesn't know any of the player's moves, Sub gets +3 score" while keeping "If the AI DOES know any of the player's moves, and those moves don't do 25%+ HP, Sub gets +3 score"

My alternative logic to handle this was just checking whether the AI had seen any of the player's moves or not, hence the new function

I don't see how it's cheating though? It was doing this only based on moves it had seen

Reply:
Maybe introducing randomness in the case of AI not knowing moves would be viable?

Reply:
Is that preferable over just not adding the score increase?

I wasn't trying to change the generic case and add "the AI randomly gets a score increase to Sub"

I was just trying to more tightly control the conditions under which Sub gets the score increase

Reply:
```
Is that preferable over just not adding the score increase if moves are unknown?
```
This could be applied to so many things and I don't think that if you don't know the moves don't increase score is a good solution to this problem because by this logic the AI should always use a damaging move turn one if it doesn't know players moves.

Reply:
> by this logic the AI should always use a damaging move turn one if it doesn't know players moves
No they just shouldn't use specifically Substitute over anything else if they don't know any of the player's moves to know that it's a sensible choice considering their HP, because Sub is uniquely extremely high ceiling low floor depending on the player's damage output in a way that no other move I can think of is

This isn't the case for the vast majority of other moves in the game, it's unique to Substitute, at least that was my idea behind implementing it

Reply:
Generally I agree with you that it isn't a good idea to use sub in this case but then there should be additional logic for when it doesn't know the moves.

Reply:
> No they just shouldn't use specifically Substitute over anything else if they don't know any of the player's moves to know that it's a sensible choice considering their HP, because Sub is uniquely extremely high ceiling low floor depending on the player's damage output in a way that no other move I can think of is

That's fair.

Reply:
Okay I'm on board for that idea, I was just going to say "no score increase if you don't know them" but that can totally work too

Is that what you meant by adding a random consideration? Like "If you don't know any moves, 50% chance for score increase. If you DO know a move and it deals <= 25%, score increase"

Reply:
I'm not sure anymore because the previous comment from you had some very valid points. Can we continue tomorrow? lol

Reply:
Absolutely lol

If it were me in a vaccuum, I would leave it without the random score increase for now, until we decide we want more comprehensive Sub handling in future.

I am also open to the random score increase, or any other ideas you might have.

Totally circle back when more awake! :D

Reply:
> If it were me in a vaccuum, I would leave it without the random score increase for now, until we decide we want more comprehensive Sub handling in future.

Ye, I agree so would just do that. The idea with checking for move none isn't bad in this case.